kernel/libtheseus/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+const _: theseus_ffi::getcwd = getcwd;
+const _: theseus_ffi::chdir = chdir;
+const _: theseus_ffi::getenv = getenv;
+const _: theseus_ffi::setenv = setenv;
+const _: theseus_ffi::unsetenv = unsetenv;
+const _: theseus_ffi::exit = exit;
+const _: theseus_ffi::getpid = getpid;
+const _: theseus_ffi::register_dtor = register_dtor;
+const _: theseus_ffi::stdin = stdin;
+const _: theseus_ffi::stdout = stdout;
+const _: theseus_ffi::stderr = stderr;
+const _: theseus_ffi::read = read;
+const _: theseus_ffi::write = write;
+const _: theseus_ffi::flush = flush;
+const _: theseus_ffi::drop_reader = drop_reader;
+const _: theseus_ffi::drop_writer = drop_writer;
+
+fn current_task() -> TaskRef {
+    task::get_my_current_task().expect("failed to get current task")
+}
+
+// Add the libtheseus:: prefix so mod_mgmt knows which crate to search.
+
+#[export_name = "libtheseus::next_u64"]
+pub extern "C" fn next_u64() -> u64 {
+    random::next_u64()
+}
+
+#[export_name = "libtheseus::getcwd"]
+pub extern "C" fn getcwd() -> FfiString {
+    current_task().get_env().lock().cwd().into()
+}
+
+#[export_name = "libtheseus::chdir"]
+pub extern "C" fn chdir(path: FfiStr<'_>) -> FfiResult<(), Error> {
+    current_task()
+        .get_env()
+        .lock()
+        .chdir(Path::new(path.into()))
+        .map_err(|e| match e {
+            environment::Error::NotADirectory => Error::NotADirectory,
+            environment::Error::NotFound => Error::NotFound,
+        })
+        .into()
+}
+
+#[export_name = "libtheseus::getenv"]
+pub extern "C" fn getenv(key: FfiStr<'_>) -> FfiOption<FfiString> {
+    current_task()
+        .get_env()
+        .lock()
+        .get(key.into())
+        .map(|s| s.to_owned().into())
+        .into()
+}
+
+#[export_name = "libtheseus::setenv"]
+pub extern "C" fn setenv(key: FfiStr<'_>, value: FfiStr<'_>) -> FfiResult<(), Error> {
+    current_task()
+        .get_env()
+        .lock()
+        .set(<&str>::from(key).to_owned(), <&str>::from(value).to_owned());
+    FfiResult::Ok(())
+}
+
+#[export_name = "libtheseus::unsetenv"]
+pub extern "C" fn unsetenv(key: FfiStr<'_>) -> FfiResult<(), Error> {
+    current_task().get_env().lock().unset(key.into());
+    FfiResult::Ok(())
+}
+
+#[export_name = "libtheseus::exit"]
+pub extern "C" fn exit(_code: i32) -> ! {
+    // TODO: Supply correct reason.
+    current_task()
+        .kill(KillReason::Requested)
+        .expect("couldn't mark task as exited");
+    task::schedule();
+    panic!("task scheduled after exiting");
+}
+
+#[export_name = "libtheseus::getpid"]
+pub extern "C" fn getpid() -> u32 {
+    task::get_my_current_task_id()
+        .try_into()
+        .expect("current task id too large")
+}
+
+#[export_name = "libtheseus::register_dtor"]
+pub unsafe extern "C" fn register_dtor(t: *mut u8, dtor: unsafe extern "C" fn(*mut u8)) {
+    unsafe { thread_local_macro::register_dtor(t, dtor) }
+}
+
+// TODO: Something better than transmutations?
```

Comment:
Bump.

Reply:
yea, can't we use a typical conversion method for this? I know there are existing fat pointer types (pointers w/ metadata) in libcore, perhaps we can leverage that (or just DIY) to restrict the interface and types that can be used to convert to/from FatPointers

Reply:
Check out this section about strict provenance and the types included therein 
https://doc.rust-lang.org/beta/core/ptr/index.html#strict-provenance

Reply:
e.g., https://doc.rust-lang.org/beta/core/primitive.pointer.html#method.to_raw_parts

Reply:
Well the problem is that the types in `core` aren't FFI-safe.

> yea, can't we use a typical conversion method for this? 

I mean the function (at least for moving from a fat pointer to a trait object) is ultimately unsafe. There's nothing guaranteeing that the implementation of the function on the other side of the FFI boundary is really returning a fat pointer corresponding to e.g. `Writer`.